File: oom-detect.c
/*****************************************************************************
 *
 *  Copyright (C) 2007-2008 Lawrence Livermore National Security, LLC.
 *  Produced at Lawrence Livermore National Laboratory.
 *  Written by Mark Grondona <mgrondona@llnl.gov>.
 *
 *  UCRL-CODE-235358
 * 
 *  This file is part of chaos-spankings, a set of spank plugins for SLURM.
 * 
 *  This is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 *  for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ****************************************************************************/

/*############################################################################
 *  $Id$
 *############################################################################
 *
 *  SLURM spank plugin to detect tasks killed by OOM killer using CHAOS
 *   kernel /proc/oomkilled file. 
 *
 *  Requires SGI Job container-based process tracking.
 *  
 *############################################################################
 */
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <dlfcn.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>


#include <job.h>
#include <slurm/spank.h>

SPANK_PLUGIN(oom-detect, 1)

typedef jid_t (*getjid_f) (pid_t pid);

static int                do_syslog = 0;

static void *             libjob =   NULL;
static getjid_f           getjid =   NULL;
static jid_t              jid =      (jid_t) -1;
static uint32_t           ntasks =   (uint32_t) -1;


int slurm_spank_init (spank_t sp, int ac, char *av[])
{
    if (!spank_remote (sp))
        return (0);

    if (ac > 0 && strcmp (av[0], "do_syslog") == 0)
        do_syslog = 1;

    if (!(libjob = dlopen ("libjob.so", RTLD_LAZY))) {
        slurm_error ("Failed to open libjob.so: %s", dlerror ());
        return (-1);
    }

    if (!(getjid = (getjid_f) dlsym (libjob, "job_getjid"))) {
        slurm_error ("Failed to resolve job_getjid in libjob: %s", 
                dlerror ());
        return (-1);
    }

    /*
     *  spank_init runs after slurm job container has been created, so
     *   now determine our jid. 
     */
    if ((jid = (*getjid) (getpid ())) == (jid_t) -1) 
        slurm_info ("Failed to get job container id");

    if (spank_get_item (sp, S_JOB_LOCAL_TASK_COUNT, &ntasks))  {
        slurm_error ("spank_get_item (S_JOB_LOCAL_TASK_COUNT) failed.");
        /* must be at least one task */
        ntasks = 1;
    }

    return (0);
}

#define OOMKILLED_FILENAME  "/proc/oomkilled"

struct oomkilled_data {
    uint64_t jobid;
    pid_t pid;
    long vmsize;
    long rss;
    char comm[16];
};

static int _parse_oomkilled (struct oomkilled_data *data, size_t size)
{
    char buf [4096];
    char *bufptr;
    char *line;
    ssize_t len;
    int count = 0;
    int fd = -1;
    int rv = -1;

    assert(data && size);

    if (access (OOMKILLED_FILENAME, R_OK) < 0) {
        goto cleanup;
    }

    if ((fd = open (OOMKILLED_FILENAME, O_RDONLY)) < 0) {
        goto cleanup;
    }

    memset(buf, '\0', sizeof (buf));
    if ((len = read (fd, buf, sizeof (buf))) < 0) {
        goto cleanup;
    }

    if (!len)
        return 0;


    line = strtok_r (buf, "\n", &bufptr);
    do {
        struct oomkilled_data *d;

        if (count >= size) {
            errno = ENOSPC;
            goto cleanup;
        }

        d = &data[count];
        memset (d, 0, sizeof (*d));
        if (sscanf (line, "%lu %d %ld %ld %15c", 
                    &d->jobid, &d->pid, &d->vmsize, &d->rss, d->comm) != 5) {
            goto cleanup;
        }
        count++;

    } while ((line = strtok_r (NULL, "\n", &bufptr))); 

    rv = count;
cleanup:
    close(fd);
    return rv;
}


int oomkilled_pids (jid_t jid, struct oomkilled_data *d, size_t len)
{
    struct oomkilled_data data [64];
    int count;
    int i;
    int index = 0;

    if ((count = _parse_oomkilled (data, 64)) < 0) {
        return -1;
    }

    for (i = 0; i < count; i++) {
        if ((jid_t) data[i].jobid == jid) {
            if (index >= len) {
                errno = ENOSPC;  
                return -1;
            }
            d[index++] = data[i];
        }
    }

    return (index);
}

static int pid_reported (pid_t pid)
{
    static pid_t pids[64];
    static int initialized = 0;
    int i = 0;

    if (!initialized) {
        memset (pids, 0, sizeof (pids));
        initialized = 1;
    }

    for (i = 0; i < 64; i++) {
        if (pids[i] == 0) {
            pids[i] = pid;
            return (0);
        }
        if (pids[i] == pid)
            return (1);
    }

    return (0);
}

static void print_oomkilled_error (struct oomkilled_data *d, int taskid)
{
    char buf [256];
    const size_t siz = sizeof (buf);
    int len = 0;

    memset (buf, 0, sizeof (buf));

    if (d->vmsize) {
        if ((len = snprintf (buf, siz, " VmSize: %ldM", d->vmsize/1024)) < 0)
            len = 0;
    }
    if (d->rss)
        len += snprintf (buf+len, siz - len, " RSS: %ldM", d->rss/1024);

    if ((len >= siz)) {
        buf [siz - 2] = '+';
        buf [siz - 1] = '\0';
    }

    if (taskid >= 0) {
        slurm_error ("task%d: [%s] terminated by OOM killer.", taskid, d->comm);
        if (d->vmsize || d->rss)
            slurm_error ("task%d:%s", taskid, buf);
    } else {
        slurm_error ("pid %ld: [%s] %s terminated by OOM killer.\n",
                d->pid, d->comm, "(task id unknown)");
        if (d->vmsize || d->rss)
            slurm_error ("pid %ld:%s", d->pid, buf);
    }
    return;
}

static void send_syslog_oom_msg (spank_t sp)
{
    uint32_t jobid;
    uint32_t stepid;
    uid_t uid;

    if ((spank_get_item (sp, S_JOB_ID, &jobid) != ESPANK_SUCCESS) ||
        (spank_get_item (sp, S_JOB_STEPID, &stepid) != ESPANK_SUCCESS) || 
        (spank_get_item (sp, S_JOB_UID, &uid) != ESPANK_SUCCESS)) {
        slurm_error ("Failed to get jobid, stepid, or uid for syslog msg.");
        return;
    }

    openlog ("slurmd", 0, LOG_USER);
    syslog (LOG_WARNING, "OOM detected: jobid=%u.%u uid=%u", jobid, stepid, uid);
    closelog ();
    slurm_verbose ("Sent OOM message via syslog for this job.");

}

int slurm_spank_task_exit (spank_t sp, int ac, char *av[])
{
    static int nexited = 0;
    struct oomkilled_data killed [16];
    int   n;
    int   i;

    if ((jid == (jid_t) -1) || (ntasks == (uint32_t) -1))
        return (0);

    ++nexited;

    /*
     *  As each task exits, report to user if any processes
     *   were terminated by OOM killer
     */
    if (!(n = oomkilled_pids (jid, killed, 16))) 
        return (0);

    for (i = 0; i < n; i++) {
        struct oomkilled_data *d = &killed[i];
        uint32_t taskid;

        if (pid_reported (d->pid))
            continue;

        spank_get_item (sp, S_JOB_PID_TO_GLOBAL_ID, d->pid, &taskid); 

        print_oomkilled_error (d, taskid);
    }

    if (nexited == ntasks) {
        if (do_syslog) 
            send_syslog_oom_msg (sp);
        /*
         *  If we got here, then we printed one or more OOM killed message 
         *   to user's  stderr. Delay a bit here to make it more likely 
         *   that the user gets the message.
         */
        sleep (2);
    }
    return (0);
}

int slurm_spank_exit (spank_t sp, int ac, char *av[])
{
    return (0);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */

File: lua/oom-detect.lua
--[==========================================================================
 *
 *  Copyright (C) 2007-2012 Lawrence Livermore National Security, LLC.
 *  Produced at Lawrence Livermore National Laboratory.
 *  Written by Mark Grondona <mgrondona@llnl.gov>.
 *
 *  UCRL-CODE-235358
 *
 *  This file is part of slurm-spank-plugins,
 *    a set of spank plugins for SLURM.
 *
 *  This is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 *  for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
-==========================================================================--]
--
--  Poor man's OOM detection plugin for SLURM.
--  
--   For each exiting task, search dmesg output for a message from
--    the OOM killer matching the exiting task's PID. If found, then
--    print an error on job's stderr and kill remaining job step tasks.
--
--   The pattern in check_oom_kill() may need to be updated for
--    different kernels. The original version in this file was for
--    RHEL6.2.
--

local posix = require 'posix'

--- Log an error with SLURM's log facility
local function log_err (...)
    SPANK.log_error (...)
end

--- Log an informational message with SLURM's log facility
local function log_info (...)
    local msg = string.format (...)
    SPANK.log_info ("oom-detect: %s", msg)
end

--- Get a spank item `item' from spank handle `spank' with name `name'
local function get_item (spank, item, name)
    local v, err = spank:get_item (item)
    if not v then
        log_err ("Unable to get %s: %s", name, err)
        return nil
    end
    return v
end

--- Create a table of job info from the spank handle `spank'
-- @spank valid spank context for a slurm job
-- Returns a job table with the following entries:
-- job
--  .jobid     SLURM job id
--  .stepid    SLURM job step id
--  .uid       user id
--  .ntasks    number of local tasks
--  .task      Table of task info
--  .task.id   Current task id
--   .pid      Current task pid
--   .status   Current task exit status
--   .exitcode Current task exit code
--   .signal   If task killed by signal, then signo
--   .coredump true if task dumped core at exit
--
local function job_table_create (spank)
    local entries = {
        jobid =  "S_JOB_ID",
        stepid = "S_JOB_STEPID",
        uid =    "S_JOB_UID",
        taskid = "S_TASK_ID",
        ntasks = "S_JOB_LOCAL_TASK_COUNT",
        pid =    "S_TASK_PID",
    }
    local job = {}

    for name,item in pairs (entries) do
        job[name] = get_item (spank, item, name)
        if not job[name] then
            return nil
        end
    end
    job.task = {
    	id = tonumber(job.taskid),
        pid = tonumber(job.pid),
    }
    job.task.status, job.task.exitcode, job.task.signal, job.task.coredump =
        spank:get_item ("S_TASK_EXIT_STATUS")

    return job
end


--- Check for OOM kill information from dmesg output `line'
function check_oom_kill (line)
    local p = "Killed process (%d+), UID %d+, %((.+)%).*:(%d+)kB.*:(%d+)kB.*:(%d+)kB"
    return string.match (line, p);
end

--- Log OOM killed task info via syslog
-- @job  job table for current job
-- @comm short command name for killed task
-- @mb   table of vm stats in MB
--
function syslog_oom_kill (job, comm, mb)
    local fmt = "OOM detected: " ..
                "jobid=%u.%u uid=%u taskid=%d ntasks=%d comm=%s " ..
                "vsz=%.1fM rss=%.1fM"
    local msg = string.format (fmt, job.jobid, job.stepid,
                               job.uid, job.task.id, job.ntasks,
                               comm, mb.vsz, mb.rss)
    posix.openlog ("slurmd")
    posix.syslog (posix.LOG_WARNING, msg)
    posix.closelog()

end

--- Log OOM kill info for job `job'
-- @job       job table for current job
-- @comm      short command name for killed task
-- @vsz       virtual address space size in Kb
-- @anon_rss  Anonymous RSS in kb
-- @file_rss  File RSS in kb
--
function log_oom_kill (job, comm, vsz, anon_rss, file_rss)
    local msg = "task%d: [%s] invoked OOM killer: vsz=%.1fM rss=%.1fM"
    local mb = { vsz = vsz / 1024,
                 rss = (anon_rss + file_rss) / 1024 }
    log_err (msg, job.taskid, comm, mb.vsz, mb.rss)
    syslog_oom_kill (job, comm, mb);
end

---
---  Kill remaining tasks in this job step after an OOM event:
-- @job job table for current job
--
function kill_all_step_tasks (job)
    local cmd = string.format ("scontrol listpids %d.%d", job.jobid, job.stepid)

    local f, err = io.popen (cmd)
    if f == nil then
        log ("%s: %s", cmd, err)
	return
    end

    local n = 0
    for line in f:lines () do
        local pid = string.match (line, "(%d+) .+")
	if pid then
	    posix.kill (pid, 9)
	    n = n + 1
	end
    end

    f:close()
end

--- Plugin hook called for each task exit event in the current job step
--
-- Check eack task exit to see if it was killed by the OOM killer, and print
--  a message to stderr and syslog if so.
--
function slurm_spank_task_exit (spank)

    local job = job_table_create (spank)

    --  If this task has been terminated by OOM killer, then exit
    --   status will be '9' (Killed). Otherwise, don't bother
    --   searching dmesg output.
    if job.task.signal ~= 9 then
        return SPANK.SUCCESS
    end

    local f, err = io.popen ("/bin/dmesg")
    if f == nil then
        log ("/bin/dmesg: %s", err)
        return SPANK.FAILURE
    end

    for line in f:lines () do
        local pid, comm, vsz, rss, file_rss = check_oom_kill (line)
        if job.task.pid == tonumber (pid) then
            log_oom_kill (job, comm, vsz, rss, file_rss)
            kill_all_step_tasks (job)
            break
        end
    end

    f:close()

    return SPANK.SUCCESS
end
